### PR TITLE
fix(managers/sbt): sbt multiple variable file lookup

### PR DESCRIPTION
**Pull request by cjtim**
~~**Working on add Unit test**~~ added

## Changes
Referece https://github.com/renovatebot/renovate/pull/16644
rewrite in `good-enough-parser`

## Context
- allow sbt manager to lookup variable from another files
- improve registry resolver (add operation `=`)
- support `var` as variable assignment
 

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [x] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/sbt/extract.spec.ts [hunk 1]
```diff
@@ -433,9 +472,308 @@ describe('modules/manager/sbt/extract', () => {
...
+                'https://repo.scala-sbt.org/scalasbt/sbt-plugin-releases',
+              ],
+            },
+          ],
+          packageFile: 'project/plugins.sbt',
+        },
+        {
+          deps: [
+            {
+              currentValue: undefined,
```
Comment (cjtim) [inline lib/modules/manager/sbt/extract.spec.ts:671]: found some bug here let me fix it

Comment (cjtim) [reply lib/modules/manager/sbt/extract.spec.ts]: fixed
[fixed by adding `var` keyword support ](https://github.com/renovatebot/renovate/pull/21622/commits/240ac8d5e909de726f432b50b5ac3d9086477740)

File: lib/modules/manager/sbt/extract.ts [hunk 2]
```diff
@@ -335,11 +385,121 @@ export function extractPackageFile(
...
+      }
+    }
   }
 
-  return { deps, packageFileVersion };
+  // Filter unique package
+  // As we merge all package to single package file
+  // Packages are counted in submodule but it's the same one
+  // by packageName and currentValue
+  const finalPackages = Object.entries(mapDepsToPackageFile).map(
```
Comment (Churro) [inline lib/modules/manager/sbt/extract.ts:488]: This is a pain to debug, please rewrite as a regular loop, filling up a new array.

Also the nesting scares me off. Better aim for a duplicate check like this:
```ts
      const depAlreadyInPkgFile = pkgFile.deps.some((item) => ...);
      if (!depAlreadyInPkgFile) {
        pkgFile.deps.push(dep);
      }
```

File: lib/modules/manager/sbt/extract.ts [hunk 3]
```diff
@@ -335,11 +385,121 @@ export function extractPackageFile(
...
+    ([packageFile, deps]) => ({
+      packageFile,
+      deps: deps.filter(
+        (val, idx, self) =>
+          idx ===
+          self.findIndex(
```
Comment (Churro) [inline lib/modules/manager/sbt/extract.ts:494]: If deduplication here happens by `packageName` and `currentValue`, I understand that a package file can include a combination of these two only once. This also means that only the first `editFile` or `fileReplacePosition` would be replaced.

File: lib/modules/manager/sbt/extract.ts [hunk 4]
```diff
@@ -335,11 +385,121 @@ export function extractPackageFile(
...
+      localVars: globalVars,
+      globalVars: {},
+      packageFile,
+    };
+    const res = extractPackageFile(content, acc);
+
+    if (res) {
+      globalVars = { ...globalVars, ...res.localVars };
+      if (res.registryUrls) {
+        registryUrls = [...registryUrls, ...new Set(res.registryUrls)];
```
Comment (Churro) [inline lib/modules/manager/sbt/extract.ts:416]: Shouldn't pushing to the `registryUrls` array be enough? Deduplication also happens later on in `extractAllPackageFiles`

File: lib/modules/manager/sbt/extract.ts [hunk 5]
```diff
@@ -335,11 +385,121 @@ export function extractPackageFile(
...
+    ...(groupPackageFileContent['.'] ?? []), // root
+  ]);
+
+  const mapDepsToPackageFile: Record<string, PackageDependency[]> = {};
+  // Start extract all package files
+  for (const packageFileContents of Object.values(groupPackageFileContent)) {
+    // Extract package file by its group
+    // local variable is share within its group
+    for (const { packageFile, content } of packageFileContents) {
+      const res = extractPackageFile(content, {
```
Comment (Churro) [inline lib/modules/manager/sbt/extract.ts:463]: This will parse at least all files of `groupPackageFileContent['project']` and `groupPackageFileContent['.']` again. It's not clear to me why this would be necessary?!

**Review by Churro — CHANGES_REQUESTED** [review]
Could you please provide a reference to the real repository you have tested this with?

Please try to reuse `resolveVariable` to match scalaVersion, packageFileVersion, etc. or at least don't repeat `ctx.localVars[varName] ?? ctx.globalVars[varName]` all over.

File: lib/modules/manager/sbt/extract.ts (hunk 5 above)
Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts:463]: This is to store global variables. before actually extract all package files

For example: in file `./build.sbt` try to resolve the variable stored in `project/Dependencies.scala`

File: lib/modules/manager/sbt/extract.ts [hunk 6]
```diff
@@ -101,13 +117,20 @@ const variableNameMatch = q
   }))
   .opt(q.op<Ctx>(':').sym('String'));
 
-const variableValueMatch = q.str<Ctx>((ctx, { value }) => {
-  ctx.vars[ctx.currentVarName!] = value;
+const variableValueMatch = q.str<Ctx>((ctx, { value, line }) => {
+  ctx.localVars[ctx.currentVarName!] = {
+    val: value,
+    sourceFile: ctx.packageFile,
+    lineIndex: line - 1,
```
[the remark this answers is not in the document]
Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts:124]: It's an index (0,1,2,3) not actual line number

File: lib/modules/manager/sbt/extract.ts [hunk 7]
```diff
@@ -335,11 +385,121 @@ export function extractPackageFile(
...
+    groupPackageFileContent[group] ??= [];
+    groupPackageFileContent[group].push({ packageFile, content });
+  }
+
+  // 1. globalVariables from project/ and root package file
+  // 2. registry from all package file
+  // 3. Project's scalaVersion - use in parseDepExpr to add suffix eg. "_2.13"
+  const { globalVars, registryUrls, scalaVersion } = prepareLoadPackageFiles([
+    ...(groupPackageFileContent['project'] ?? []), // in project/ folder
+    ...(groupPackageFileContent['.'] ?? []), // root
```
[the remark this answers is not in the document]
Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts:454]: Ah you're right in original Pullrequest i have 3 difference variable
```js
  variables?: Variables; // within single file 
  localVariables?: Variables; // variable within subfolder
  globalVariables?: Variables;
```

File: lib/modules/manager/sbt/extract.ts [hunk 8]
```diff
@@ -289,13 +323,21 @@ const query = q.tree<Ctx>({
   postHandler: registryUrlHandler,
 });
 
+// Extract 1 file
 export function extractPackageFile(
```
Comment (zharinov) [inline lib/modules/manager/sbt/extract.ts:327]: `extractPackageFile` is reserved for the upper-level API usage, please rename it to avoid confusion

File: lib/modules/manager/sbt/extract.ts [hunk 9]
```diff
@@ -289,13 +323,21 @@ const query = q.tree<Ctx>({
...
   content: string,
-  packageFile: string
-): PackageFileContent | null {
+  {
+    packageFile,
+    registryUrls,
+    localVars,
+    globalVars,
+    scalaVersion,
+  }: PackageFile & ParseOptions
```
Comment (zharinov) [inline lib/modules/manager/sbt/extract.ts:335]: Please don't use type intersection, instead add the missing fields to the `ParseOptions`

**Review by zharinov — CHANGES_REQUESTED** [review]

File: lib/modules/manager/sbt/__snapshots__/extract.spec.ts.snap [hunk 10]
```diff
@@ -105,7 +127,19 @@ exports[`modules/manager/sbt/extract extractPackageFile() extract deps from nati
       "variableName": "abcVersion",
     },
   ],
-  "packageFileVersion": undefined,
+  "globalVars": {},
+  "localVars": {
```
Comment (viceice) [inline lib/modules/manager/sbt/__snapshots__/extract.spec.ts.snap:131]: this should not be on results object

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/sbt/extract.ts (hunk 9 above)
Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts:334]: done

File: lib/modules/manager/sbt/extract.ts (hunk 8 above)
Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts]: renamed

File: lib/modules/manager/sbt/extract.ts (hunk 4 above)
Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts]: i fixed this

File: lib/modules/manager/sbt/extract.ts [hunk 11]
```diff
@@ -373,9 +372,7 @@ export function extractPackageFile(
       globalVars,
       localVars,
       deps: [],
-      registryUrls: [
-        ...new Set([REGISTRY_URLS.mavenCentral, ...(registryUrls ?? [])]),
-      ],
+      registryUrls: [REGISTRY_URLS.mavenCentral, ...(registryUrls ?? [])],
```
Comment (viceice) [inline lib/modules/manager/sbt/extract.ts:375]: why this change?

File: lib/modules/manager/sbt/extract.ts [hunk 12]
```diff
@@ -410,12 +407,12 @@ function prepareLoadPackageFiles(
...
       packageFile,
     };
-    const res = extractPackageFile(content, acc);
+    const res = extractDependency(content, acc);
 
     if (res) {
       globalVars = { ...globalVars, ...res.localVars };
       if (res.registryUrls) {
-        registryUrls = [...registryUrls, ...new Set(res.registryUrls)];
+        registryUrls = [...registryUrls, ...res.registryUrls];
```
Comment (viceice) [inline lib/modules/manager/sbt/extract.ts:415]: see other comment.

Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts:415]: I moved to `extractAllPackageFiles` so it invokes new Set only 1 time before return final result

File: lib/modules/manager/sbt/types.ts [hunk 13]
```diff
@@ -0,0 +1,19 @@
...
+};
+export type Variables = Record<string, VariableContext>;
+
+export type GroupFilenameContent = Record<
+  string,
+  { packageFile: string; content: string }[]
+>;
+
+export interface ParseOptions {
+  packageFile?: string;
```
[the remark this answers is not in the document]
Comment (cjtim) [reply lib/modules/manager/sbt/types.ts]: removed

File: lib/modules/manager/sbt/extract.ts (hunk 3 above)
Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts]: https://github.com/renovatebot/renovate/pull/21622/files#diff-a550300890b64a89e53f84f51ad5d5a439e7fb8c75159f8ea4bbf0641e6729a1R471-R477

I did some changes to simplify this

File: lib/modules/manager/sbt/extract.ts [hunk 14]
```diff
@@ -309,10 +349,15 @@ export function extractPackageFile(
         currentValue: sbtVersion,
         replaceString: matchString,
         extractVersion: '^v(?<version>\\S+)',
+        editFile: packageFile,
```
[the remark this answers is not in the document]
Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts]: i removed this line

File: lib/modules/manager/sbt/update.ts [hunk 15]
```diff
@@ -31,3 +34,31 @@ export function bumpPackageVersion(
 
   return { bumpedContent };
 }
+
+export function updateDependency({
```
[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/manager/sbt/update.ts:38]: and also set packageFile to the file, where the replacement should be done

File: lib/modules/manager/sbt/extract.ts [hunk 16]
```diff
@@ -335,11 +383,107 @@ export function extractPackageFile(
...
+      const res = extractDependency(content, {
+        registryUrls,
+        deps: [],
+        packageFile,
+        scalaVersion,
+        localVars: {},
+        globalVars,
+      });
+      for (const dep of res?.deps ?? []) {
+        const variableSourceFile = dep?.editFile ?? packageFile;
```
Comment (viceice) [inline lib/modules/manager/sbt/extract.ts:468]: remove all properties which are not known to the dep interface before returning them, like editFile and variables

**Review by viceice — CHANGES_REQUESTED** [review]

**Review by Churro — CHANGES_REQUESTED** [review]
See unresolved (but due to force-push outdated) remarks.

**Comment by cjtim** [comment]
Have some questions there's 2 function in `extract.ts` file
* `extractDependency` used internally by `extractAllPackageFiles` and test case
* `extractAllPackageFiles` called by the manager module and test case

Should I use `extractDependency` in the test case or not?

File: lib/modules/manager/sbt/extract.ts (hunk 16 above)
Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts:464]: deleted

File: lib/modules/manager/sbt/types.ts [hunk 17]
```diff
@@ -0,0 +1,19 @@
+export type VariableContext = {
+  val: string;
+  sourceFile: string;
+  lineIndex: number;
```
[the remark this answers is not in the document]
Comment (cjtim) [reply lib/modules/manager/sbt/types.ts]: replaced

File: lib/modules/manager/sbt/update.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (cjtim) [reply lib/modules/manager/sbt/update.ts:38]: But that wouldn't cover the case where the same version is present in the same file right?

**Comment by viceice** [comment]
> Have some questions there's 2 function in `extract.ts` file
> * `extractDependency` used internally by `extractAllPackageFiles` and test case
> * `extractAllPackageFiles` called by the manager module and test case
> 
> Should I use `extractDependency` in the test case or not?

if you can reach 100% code coverage with `extractAllPackageFiles` then you don't need to test the other function explicit.

File: lib/modules/manager/sbt/update.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (viceice) [reply lib/modules/manager/sbt/update.ts:38]: why it doesn't cover? i don't understand the issue.

renovate will use the `packageFile` later to auto replace the version in that file. if multiple deps are referring to the same variable, group them, so it's a single PR for a single version.

@rarkins i think it's time for an `srcFile` property on the dependency, so auto replace can use that optional fileName to replace the version in the right place.that would probably also help the other managers (maven, gradle) which currently use a custom update function because of that.

File: lib/modules/manager/sbt/types.ts [hunk 18]
```diff
@@ -0,0 +1,19 @@
+export type VariableContext = {
+  val: string;
+  sourceFile: string;
```
[the remark this answers is not in the document]
Comment (cjtim) [reply lib/modules/manager/sbt/types.ts]: renamed

File: lib/modules/manager/sbt/types.ts [hunk 19]
```diff
@@ -0,0 +1,19 @@
+export type VariableContext = {
+  val: string;
```
[the remark this answers is not in the document]
Comment (cjtim) [reply lib/modules/manager/sbt/types.ts]: renamed

File: lib/modules/manager/sbt/extract.ts [hunk 20]
```diff
@@ -210,6 +238,10 @@ function depHandler(ctx: Ctx): Ctx {
   if (variableName) {
     dep.groupName = variableName;
     dep.variableName = variableName;
+    if (currentValueInfo) {
+      dep.fileReplacePosition = currentValueInfo.lineIndex;
```
[the remark this answers is not in the document]
Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts]: moved to managerData

File: lib/modules/manager/sbt/extract.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts]: moved to lineNumber

File: lib/modules/manager/sbt/update.ts (hunk 15 above)
[the remark this answers is not in the document]
Comment (cjtim) [reply lib/modules/manager/sbt/update.ts:39]: I tested
```
 INFO: Failed to parse branchContent - rebasing (repository=username/dummpyrepo, packageFile=main/build.sbt, branch=renovate/com.netflix.archaius-archaius-core-0.x)
       "manager": "sbt"
DEBUG: Rebasing branch after deps list has changed (repository=username/dummpyrepo, packageFile=main/build.sbt, branch=renovate/com.netflix.archaius-archaius-core-0.x)
       "depName": "com.netflix.archaius:archaius-core"
DEBUG: manager.getUpdatedPackageFiles() reuseExistingBranch=false (repository=username/dummpyrepo, branch=renovate/com.netflix.archaius-archaius-core-0.x)
DEBUG: Starting search at index 8024 (repository=username/dummpyrepo, packageFile=main/build.sbt, branch=renovate/com.netflix.archaius-archaius-core-0.x)
       "depName": "com.netflix.archaius:archaius-core"
DEBUG: Found match at index 8024 (repository=username/dummpyrepo, packageFile=main/build.sbt, branch=renovate/com.netflix.archaius-archaius-core-0.x)
       "depName": "com.netflix.archaius:archaius-core"
DEBUG: Could not extract main/build.sbt (repository=username/dummpyrepo, branch=renovate/com.netflix.archaius-archaius-core-0.x)
 WARN: Error updating branch: update failure
```

`doAutoReplace` function does not work since it calls `confirmIfDepUpdated` which refers to `extractPackageFile` function which I replaced by `extractAllPackageFiles`

[the remark this answers is not in the document]
Comment (Churro) [reply lib/modules/manager/sbt/update.ts:39]: Then it would be great to have `extractPackageFile` back. 

`extractAllPackageFiles` typically calls `extractPackageFile`, see npm or batect mangers. `extractAllPackageFiles` should do the same but on `n` packageFiles, instead of just 1 as in `extractPackageFile`.

File: lib/modules/manager/sbt/extract.ts [hunk 21]
```diff
@@ -323,23 +373,123 @@ export function extractPackageFile(
...
+    ...(groupPackageFileContent['.'] ?? []), // root
+  ]);
+
+  const mapDepsToPackageFile: Record<
+    string,
+    PackageDependency<SbtManagerData>[]
+  > = {};
+  // Start extract all package files
+  for (const packageFileContents of Object.values(groupPackageFileContent)) {
+    for (const { packageFile, content } of packageFileContents) {
```
Comment (Churro) [inline lib/modules/manager/sbt/extract.ts:461]: At this point you may implement `extractPackageFile(content, file, config)`

File: lib/modules/manager/sbt/extract.ts [hunk 22]
```diff
@@ -323,23 +373,123 @@ export function extractPackageFile(
...
+      localVars,
       deps: [],
-      registryUrls: [REGISTRY_URLS.mavenCentral],
+      registryUrls,
+      packageFile,
+      scalaVersion,
     });
   } catch (err) /* istanbul ignore next */ {
-    logger.debug({ err, packageFile }, 'Sbt parsing error');
+    logger.warn({ err, packageFile }, 'Sbt parsing error');
```
Comment (Churro) [inline lib/modules/manager/sbt/extract.ts:384]:
```suggestion
    logger.debug({ err, packageFile }, 'Sbt parsing error');
```

File: lib/modules/manager/sbt/extract.ts [hunk 23]
```diff
@@ -267,7 +309,7 @@ const addResolverMatch = q.sym<Ctx>('resolvers').alt(
 
 function registryUrlHandler(ctx: Ctx): Ctx {
```
Comment (Churro) [inline lib/modules/manager/sbt/extract.ts:310]: Why don't you apply the fallback to maven central **once** instead of passing it all over? E.g.:

```ts
function registryUrlHandler(ctx: Ctx): Ctx {
  ctx.registryUrls = [...new Set(ctx.registryUrls)];
  if (!ctx.registryUrls.length) {
    ctx.registryUrls.push(REGISTRY_URLS.mavenCentral);
  }

  for (const dep of ctx.deps) {
    dep.registryUrls = [...ctx.registryUrls];
    if (dep.depType === 'plugin') {
      dep.registryUrls.push(SBT_PLUGINS_REPO);
    }
  }
  return ctx;
}
```

File: lib/modules/manager/sbt/extract.ts [hunk 24]
```diff
@@ -323,23 +373,123 @@ export function extractPackageFile(
 
   try {
     parsedResult = scala.query(content, query, {
```
Comment (Churro) [inline lib/modules/manager/sbt/extract.ts:375]: Why do you re-initialize `Ctx` over and over for each dependency? It would be a lot simpler do initialize it just once, and use it to access `globalVars`, `registryUrls`, `deps`, etc. instead of passing them as arguments. You can still have temp variables in there. See gradle manager for ref.

File: lib/modules/manager/sbt/extract.ts [hunk 25]
```diff
@@ -323,23 +373,123 @@ export function extractPackageFile(
...
+function prepareLoadPackageFiles(
+  packageFilesContent: { packageFile: string; content: string }[]
+): Pick<Ctx, 'globalVars' | 'registryUrls' | 'scalaVersion'> {
+  // Return variable
+  let globalVars: Ctx['globalVars'] = {};
+  let registryUrls: Ctx['registryUrls'] = [REGISTRY_URLS.mavenCentral];
+  let scalaVersion: Ctx['scalaVersion'] = undefined;
+
+  for (const { packageFile, content } of packageFilesContent) {
+    const res = extractDependency(content, {
```
Comment (Churro) [inline lib/modules/manager/sbt/extract.ts:404]: This is very inefficient because with `scala.query` in `extractDependency`, you also extract dependencies which you then throw away here. I'd rather make two separate queries, so that you don't match absolutely **everything** twice.

File: lib/modules/manager/sbt/extract.ts [hunk 26]
```diff
@@ -323,23 +373,123 @@ export function extractPackageFile(
...
+  ]);
+
+  const mapDepsToPackageFile: Record<
+    string,
+    PackageDependency<SbtManagerData>[]
+  > = {};
+  // Start extract all package files
+  for (const packageFileContents of Object.values(groupPackageFileContent)) {
+    for (const { packageFile, content } of packageFileContents) {
+      const res = extractDependency(content, {
```
Comment (Churro) [inline lib/modules/manager/sbt/extract.ts:462]: See comment above. This is very inefficient because with `scala.query` in `extractDependency`, you also extract `globalVars`, `localVars`, `registryUrls`, again which you then throw away here. I'd rather make two separate queries, so that you don't match absolutely **everything** twice.

**Review by Churro — CHANGES_REQUESTED** [review]

File: lib/modules/manager/sbt/extract.ts (hunk 25 above)
Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts:404]: That is exactly what I wanted. The first iteration is to store variables without this it wouldn't be able to resolve `Versions.akka` in this file https://github.com/renovatebot/renovate/pull/21622/files#diff-09d027666d4c679177a665d5920a343d2b1ef0c1cad4a4822a649fb991076910R35

You can take a look on `simple-project` without do it twice it would have a race condition on it

**Comment by cjtim** [comment]
Have some bug which i'm not sure how to solve. please help guide

The title is updated but not the branch itself. You can see prBody and prTitle is updated but new version is not push to the branch

Screenshot 2566-05-30 at 10 37 24

File: lib/modules/manager/sbt/extract.ts (hunk 21 above)
Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts:461]: https://github.com/renovatebot/renovate/pull/21622#discussion_r1189771459

I think we need to avoid using reserve function name

File: lib/modules/manager/sbt/extract.ts (hunk 25 above)
Comment (cjtim) [reply lib/modules/manager/sbt/extract.ts:404]: I still don't have much idea to resolve the variable. Without populating it in the first loop. 
It's like we can do it all without 2 loops but we'll be missing some variables.

Comment (Churro) [reply lib/modules/manager/sbt/extract.ts:404]: Probably that was a misunderstanding but I didn't mean there should be only one loop. The idea was rather to have to distinct `scala.query` operations that each encompass different queries. In order words, split querying global vars from the rest.

**Comment by rarkins** [comment]
Closing due to inactivity. PR is still welcome any time in future
